src/monitors/R3Monitor.v3

```diff
@@ -0,0 +1,728 @@
+		var dst_offset = Values.v_u(frame_accessor.getOperand(-2));
+		var val = Values.v_u(frame_accessor.getOperand(-1));
+		var size = Values.v_u(frame_accessor.getTopOfStack());
+		handler.onMemoryFill(mem_index, dst_offset, u8.view(val), size);
+		return Resumption.Continue;
+	}
+}
+
+private type CallStackEntry(func_id: int, kind: CallKind);
+
+type WasmEvent {
+	// function calls by the host
+	case FuncEntry(idx: int, func_name: string, params: Array<Value>);
+	case FuncEntryTable(idx: int, table_idx: int, table_name: string, params: Array<Value>);
+
+	// observed memory updates made by the host
+	case Load(idx: int, offset: u64, data: Array<u8>);
+	case MemGrow(idx: int, amount: int);
+
+	// observed table updates made by the host
+	case TableGet(table_idx: int, table_name: string, entry_idx: int, func_idx: int, func_name: string);
+	case TableGrow(table_idx: int, table_name: string, amount: u32);
+
+	// observed global updates made by the host
+	case GlobalGet(idx: int, value: Value);
+
+	// imported function call/returns made by the module
+	case ImportCall(idx: int);
+	case ImportReturn(results: Array<Value>);
+
+	// initial value of imported globals
+	case ImportGlobal(idx: int, initial: Value);
+
+	def render(sb: StringBuilder) {
+		match (this) {
+			FuncEntry(idx, func_name, params) => {
+				sb.put2("EC;%d;%s;", idx, func_name);
+				renderParams(sb, params);
+			}
+			FuncEntryTable(idx, table_idx, table_name, params) => {
+				sb.put3("TC;%d;%s;%d;", idx, table_name, table_idx);
+				renderParams(sb, params);
+			}
+			Load(idx, offset, data) => {
+				sb.put2("L;%d;%d;", idx, offset);
+				renderBytes(sb, data);
+			}
+			MemGrow(idx, amount) => {
+				sb.put2("MG;%d;%d", idx, amount);
+			}
+			TableGet(table_idx, table_name, idx, func_idx, func_name) => {
+				sb.put3("T;%d;%s;%d;", table_idx, table_name, idx);
+				sb.put2("%d;%s", func_idx, func_name);
+			}
+			TableGrow(table_idx, table_name, amount) => {
+				sb.put3("TG;%d;%s;%d", table_idx, table_name, amount);
+			}
+			GlobalGet(idx, value) => {
+				sb.put1("G;%d;", idx);
+				renderValue(value, sb);
+			}
+			ImportCall(idx) => {
+				sb.put1("IC;%d", idx);
+			}
+			ImportReturn(results) => {
+				sb.puts("IR;0;");
+				renderParams(sb, results);
+			}
+			ImportGlobal(idx, initial) => {
+				sb.put1("IG;%d;", idx);
+				renderValue(initial, sb);
+			}
+		}
+		sb.puts("\n");
+	}
+
+	// helper method to print values for globals related events
+	def renderValue(val: Value, sb: StringBuilder) {
+		if (val == Values.F32_nan || val == Values.F64_nan) {
+			sb.puts("NaN");
+		} else if (val == Values.F32_infinity || val == Values.F64_infinity) {
+			sb.puts("Infinity");
+		} else if (val == Values.F32_minus_infinity || val == Values.F64_minus_infinity) {
+			sb.puts("-Infinity");
+		} else {
+			match (val) {
+				I32(val) => sb.put1("%d", i32.view(val));
+				I64(val) => sb.put1("%d", i64.view(val));
+				F32(val) => sb.put1("%s", Floats.displayFromBits32(val));
```

Comment:
I prefer using something like "0x<hex_from_float>" in the trace to represent floats, it produces the most accurate representation. @doehyunbaek it should be easy to implement in the backend, right? we can also support both formats.

For this PR, I'll go back to using the hex format then, and modify the tests accordingly.

Reply:
I think it wouldnt be hard to support hex format in the backend.

Not sure Wasabi based frontend can output in hex format, let's go for each outputting its preferred format(Wasabi rounding, Wizard hex) and later decide if we can ditch the rounding format of current Wasabi implementation

Reply:
Wasabi is probably using JS to render the value. But AFAICT there is no need to involve floating point types for memory, only for globals. It's best to avoid anything that doesn't preserve the exact floating point bits, so hex seems like the only solution going forward.

For the Wasabi frontend, one way to get the exact bits for globals is to would be to use a JS typed array. But that risks going through a JS value. Maybe better to have a Wasm module that imports a floating-point typed global, reads it, reinterprets it, and returns it as an `i64`.